Use @test so that TestSetExtensions can intercept

oxinabox opened the pull request:
I need to sleep now, here is what I have so far.
I've only touched strings so far.

![image](https://user-images.githubusercontent.com/5127634/32786273-6fa1079c-c98e-11e7-84a2-ed558cae7a91.png)

See example in `/demo/demo.jl` (maybe not the best location for it).

So as not to need to added `TestSetExtensions` or `DeepDiffs` as a dependency.
It seems best to use `@test` to execute the tests.
Then `TestSetExtensions` will just intercept the test message response.

Problem is this doesn't allow the chance to let the user re-generate the reference after seeing the 
diff.

My resolution was to not allow the user to re-generate references at all.
They can just delete the reference files if they want them regenerated.
I think that makes sense from a usage perspective: "Delete references that no longer match the expectations."
The continuation from there would be to automatically (without user input) generate references for when files are not found (while displaying a warning).
I don't see why not to do that.
(but I might be a bit too tired to reason right).

( CC @ssfrr )

Evizero commented:
I like the diff, but I am reserved about this. The interactive updating is my main use case for this package.

oxinabox commented:
I'll have a think about the other options when I am more awake.
There might be some golden path, that has the best of both worlds,
by (e.g.) deferring handling the failures til after the testsets errors are displayed.
(Well one way is to just use DeepDiff directly)

Evizero commented:
I'd be fine with that dependency. Its very light weight

oxinabox commented:
other thing is though,
throwing `error`s and using `@test true` and `@test false`  and directly writing own outputs,
feels like an abuse of the `Base.Test` mechanics,
that going to have its own set of problems.

(zzzz... now. thanks for looking at it)

Evizero commented:
Yea I agree. I am also not the biggest fan of this "hack".

Maybe there is a clean way to do it with the internals of `Base.Test`? I never actually looked at what `@test` does.

Evizero commented:
Thanks for tackling this, way the way. Diffs would be a really nice addition

oxinabox commented:
I think a custom AbstractTestset  might be able to do it.
https://github.com/JuliaLang/julia/blob/master/stdlib/Test/docs/src/index.md#creating-custom-abstracttestset-types
Though it might have to reorder its own position in the stack.

ssfrr commented:
That's how TestSetExtensions does it (custom `AbstractTestSet` subtype). I set it up so that you should be able to wrap a different testset type, e.g. `ExtendedTestSet(ReferenceTestSet)` and get the `ExtendedTestSet` display stuff while still allowing the wrapped `ReferenceTestSet` type to do its thing.

That sort of `TestSet` composition is one of the areas that might cause some trickiness, but that's how it should work in principle. :)

oxinabox commented:
So I have tried to understand how testsets are meant to work.
So I think the "intended" way to do this kind of thing is to introduce a new TestSet type.
To process a new Result type.

In this case the new Result types would be `ReferenceMissing` and `ReferenceDisagreement`.
The new testset type would handle those by allowing the user to regenerate them.
It would also git its parent testset to actually process the associated failures to match.

I have a working prototype here,
for `ReferenceMissing` but the same goes for `ReferenceDisagreement`.

Big downside I see to this is that it is absolutely required to run reference tests inside a `ReferenceTestset`.
My gut tells me the more sensible situtation is that you only get interactive options inside a `ReferenceTestset` but that they give `Pass`es or `Fail`ures in other testsets.
But that doesn't seem easy to do, without hackily hiding extra information inside a normal `Pass` or `Fail` result.

Evizero commented:
This seems to be the right direction in terms of proper tests. The need for specifying a custom TestSet type seems unfortunate though. Browsing through the `Base.Test`code, shouldn't it also be possible to just overload `record(::DefaultTestSet, ::MissingFile)` ?

oxinabox commented:
@Evizero I've found some time to take another swing.
As you said I added a generic method to all testsets's record.
Then I thought some more and realised I am not really benifitting from subtyping `Result` and using `record` at all.
So I changed it so it just called `record` on its fakedup inner Result.


To make that result less bad we would need to capture the actual `Expr`  and the `@__LINE__` on the `@test_reference` macro call, and feed that down through all the functions.
And also scrub the back-traces to remove all between function invocations til we get to the file where `@test_reference` was actually called  (That is actually how `Base.Test` does it, so if it is good for them it is good for us).


Some thing to note:
`MissingFile` and `MismatchedFile` are almost identical. 
    - Possibly they could be one type with just a field saying if they are different. 
    - I feel semantically they are very different though. They might want very different behavior, e.g. Maybe `MissingFile` should just create the files and then give a warning. That would be a future change though, so maybe I am adding unnesc. flexibility now.

Evizero commented on demo/demo.jl:
```diff
@@ -0,0 +1,33 @@
+using ReferenceTests
+
+
+@testset  ExtendedTestSet "all the tests" begin
```
i am guessing this is just a leftover?

oxinabox replied:
No?
Run the demo, and it shows that the `ExtendedTestSet` works, and does it's core functionality of displaying Diffs.

Evizero replied:
oh, I see. it's from `TestSetExtensions`. So to be clear, in case one doesn't care about the diff display, the functionality here itself doesn't require using `ExtendedTestSet` right?

oxinabox replied:
exactly.

Evizero commented:
I like the way this looks. does it run? how would the package behaviour be different?

> MissingFile and MismatchedFile are almost identical.

Right. I agree though that we should just add both anyway

c42f commented:
i've been following this one in relation to the harvesting of nonstandard `Pass`/`Fail` `Result` types. I do think we need something better in `Test` for this. I hacked around it while implementing `@test_logs` (which has just been merged into `Test` in initial form), but I made a small mess there for the sake of expediency in getting the logging stuff merged.

oxinabox commented:
What is the expected behavour from `@test_reference "references/string1.txt" [foo * "bar"]` 
More generally what is the expected behavior  `@test_reference(path, xs::Vector)` 

Is it something like `@testreference(path, join(xs, "\n"))` ?
and then a general fallback for `@testreference(path, x::Any) = @testreference(path, repr(x))` ?


The current build fail is because of got rid of the definitions for `@test_reference(path, ::Vector{<:AbstractString})`. 
Because that was doing it backwards -- it was splitting strings to then handle them as arrays.
(which introduced a bunch of weirdness like what had to be fixed in #2)

Where as now for making Diffs work, we want to always just be comparing multiline strings.

oxinabox commented:
Ok, assuming this passes CI I am done messing with this.

There is still some old code that does images that could be refactored,
but as a non-image person, I don't have much interest or expertise in it.
(though I must say ImagesInTerminal is super nifty)

oxinabox commented on test/runtests.jl:
```diff
@@ -34,7 +68,7 @@ end
 @testset "string as txt" begin
     foo = "foo"
-    @test_reference "references/string1.txt" foo * "bar"
+    @show @macroexpand @test_reference "references/string1.txt" foo * "bar"
```
This line needs to be deleted.

oxinabox commented:
propagating test coverage  through into the spawned process might be possible.
I think that is where the 6.5% went

oxinabox commented:
I've been using this today in my tests for DataDepGenerators.jl,
with ExtendedTestSets so diffs do show up.
and it is really nice.

(DataDeps generators is purely tested by reference tests, because I am lazy, and this stuff is hard to get in and unit test properly)

Evizero commented:
Right now it breaks my main usecase of using ImageInTerminal.

e.g. if I delete a the `camera.txt` reference file i now get this output:

```
images as txt using ImageInTerminal: Test Failed
  Expression: @test_reference "/home/csto/.julia/v0.6/ReferenceTests/test/references/camera.txt" [missing] == actual
   Evaluated: "" == "\e[0m\e[38;5;236;48;5;239m▀\e[38;5;239;48;5;247m▀\e[38;5;240;48;5;248m▀\e[38;5;240;48;5;246m▀\e[38;5;240;48;5;246m▀\e[38;5;240;48;5;248m▀
\e[38;5;240;48;5;247m▀\e[38;5;239;48;5;246m▀\e[38;5;235;48;5;239m▀\e[0m\n\e[0m\e[38;5;239;48;5;239m▀\e[38;5;247;48;5;244m▀\e[38;5;244;48;5;235m▀\e[38;5;239;48
;5;234m▀\e[38;5;240;48;5;238m▀\e[38;5;245;48;5;244m▀\e[38;5;247;48;5;247m▀\e[38;5;246;48;5;246m▀\e[38;5;239;48;5;239m▀\e[0m\n\e[0m\e[38;5;239;48;5;238m▀\e[38;
5;243;48;5;239m▀\e[38;5;234;48;5;233m▀\e[38;5;234;48;5;236m▀\e[38;5;238;48;5;242m▀\e[38;5;244;48;5;243m▀\e[38;5;247;48;5;244m▀\e[38;5;246;48;5;244m▀\e[38;5;23
9;48;5;237m▀\e[0m\n\e[0m\e[38;5;238;48;5;238m▀\e[38;5;238;48;5;240m▀\e[38;5;233;48;5;235m▀\e[38;5;238;48;5;241m▀\e[38;5;243;48;5;243m▀\e[38;5;243;48;5;243m▀\e
[38;5;242;48;5;243m▀\e[38;5;242;48;5;243m▀\e[38;5;237;48;5;237m▀\e[0m\n\e[0m\e[38;5;235m▀\e[38;5;236m▀\e[38;5;235m▀\e[38;5;237m▀\e[38;5;238m▀\e[38;5;238m▀\e[3
8;5;237m▀\e[38;5;237m▀\e[38;5;235m▀\e[0m"
Stacktrace:
```

Evizero commented:
Maybe we need a layered approach to this. Given that your use-case is somewhat different and probably more widespread/useful, it might make sense to have a slimmed down version of this in its own package. This would also solve the dependency dilemma

oxinabox commented:
hmm maybe. We could work around images, of course.
e,g, pass down to the methods more info, so it can dispatch differently for images vs text vs numbers

oxinabox commented:
We should attempt to work this out.
This is maximally useful for DataDepsGenerators.jl.
And I have a GSOC proposal open for that,  if I get/accept a student, then it will be useful for them.

oxinabox commented:
@SebastinSanty is said GSOC student BTW, @Evizero.

@SebastinSanty would making these Diffs work in DataDepsGeneratrors.jl,
make your life easier, each time you regenerate the tests?
Or is the current status no big issue

SebastinSanty commented:
@oxinabox Yes, it would definitely help as I will get an idea as to what is getting replaced. Is it because of the code changes I made or because the content of the reference actually changed.

oxinabox commented:
Ok, I will attempt to get this branch working for both the simple text use case and for @Evizero's fancy image use case.

SebastinSanty commented:
I am all up for this coming in. Its getting increasingly difficult for me to detect how the references have changed over time.

oxinabox commented:
Anyone know what I was doing with my changes in `test/runtests.jl`?

I've started this PR from scratch, without touching the base, testing infrastructure directly (for now)
and I can't workout if what I was doing there mattered at all.

oxinabox commented:
Close in favor of #7
